File: anvil-recyclerview-v7/src/main/java/trikita/anvil/RenderableRecyclerViewAdapter.java
package trikita.anvil;

import android.view.ViewGroup;
import android.widget.FrameLayout;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public abstract class RenderableRecyclerViewAdapter
        extends RecyclerView.Adapter<RenderableRecyclerViewAdapter.MountHolder> {

    public static <T> RenderableRecyclerViewAdapter withItems(final List<T> items,
            final RenderableAdapter.Item<T> r) {
        return new RenderableRecyclerViewAdapter() {
            {
                setHasStableIds(false);
            }
            public int getItemCount() {
                return items.size();
            }
            public int getItemViewType(int pos) {
                Object item = items.get(pos);
                return item == null ? 0 : item.getClass().hashCode();
            }
            public void view(RecyclerView.ViewHolder holder) {
                int i = holder.getLayoutPosition();
                r.view(i, items.get(i));
            }
        };
    }

    @Override
    public MountHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        FrameLayout root = new FrameLayout(parent.getContext());
        root.setLayoutParams(new ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT));
        return new MountHolder(root);
    }

    @Override
    public void onBindViewHolder(final MountHolder h, int position) {
        if (h.mount == null) {
            h.mount = new Anvil.Mount((ViewGroup) h.itemView, new Anvil.Renderable() {
                public void view() {
                    RenderableRecyclerViewAdapter.this.view(h);
                }
            });
            Anvil.render(h.mount);
        } else {
            Anvil.render(h.mount);
        }
    }

    @Override
    public long getItemId(int pos) {
        return pos; // just a most common implementation
    }

    public static class MountHolder extends RecyclerView.ViewHolder {
        private Anvil.Mount mount;

        public MountHolder(ViewGroup itemView) {
            super(itemView);
        }
    }

    public abstract void view(RecyclerView.ViewHolder holder);
}

File: anvil-recyclerview-v7/src/main/kotlin/dev/inkremental/dsl/androidx/recyclerview/RenderableRecyclerViewAdapter.kt
package dev.inkremental.dsl.androidx.recyclerview

import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.recyclerview.widget.RecyclerView
import dev.inkremental.Inkremental.Mount
import dev.inkremental.Inkremental.render
import dev.inkremental.RenderableAdapter

abstract class RenderableRecyclerViewAdapter<T>(val fillHeight: Boolean = false) : RecyclerView.Adapter<RenderableRecyclerViewAdapter.MountHolder>() {

    open lateinit var items: List<T>

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MountHolder {
        val root = FrameLayout(parent.context)
        root.layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                if (fillHeight) ViewGroup.LayoutParams.MATCH_PARENT else ViewGroup.LayoutParams.WRAP_CONTENT)
        return MountHolder(root)
    }

    override fun onBindViewHolder(h: MountHolder, position: Int) {
        if (h.mount == null) {
            h.mount = Mount(h.itemView as ViewGroup) { view(h) }
            render(h.mount!!)
        } else {
            render(h.mount!!)
        }
    }

    override fun getItemId(pos: Int): Long {
        return pos.toLong() // just a most common implementation
    }

    class MountHolder(itemView: ViewGroup) : RecyclerView.ViewHolder(itemView) {
        var mount: Mount? = null
    }

    abstract fun view(holder: RecyclerView.ViewHolder?)

    companion object {
        fun <T> withItems(i: List<T>,
                          fillHeight: Boolean = false,
                          r: (index: Int, item: T) -> Unit): RenderableRecyclerViewAdapter<T> {
            return object : RenderableRecyclerViewAdapter<T>(fillHeight) {

                override var items = i

                override fun getItemCount(): Int {
                    return items.size
                }

                override fun getItemViewType(pos: Int): Int {
                    val item: Any? = items[pos]
                    return item?.javaClass?.hashCode() ?: 0
                }

                override fun view(holder: RecyclerView.ViewHolder?) {
                    holder?.let {
                        val i = it.layoutPosition
                        r(i, items[i])
                    }
                }

                init {
                    setHasStableIds(false)
                }
            }
        }
    }
}
